test/helpers/fuzzing.js

```diff
@@ -0,0 +1,242 @@
+      return cb(new Error(`missing keys: ${keys}`))
+    })
+    readStream.on('error', cb)
+    readStream.on('data', function (nodes) {
+      if (!nodes) return
+      var key = nodes[0].key
+      var values = nodes.map(node => node.value)
+      t.same(values, expectedWrites[key])
+      delete expectedWrites[key]
+    })
+  })
+}
+
+function generateFailingTest (name, dbCount, writesPerReplication, writeOps, cb) {
+  writeOps.push(err => {
+    t.error(err)
+    t.end()
+  }).toString()
+  fs.writeFile(p.join(__dirname, '..', name + '.js'), prettier.format(`
+    var tape = require('tape')  
+
+    var run = require('./helpers/run')
+    var put = require('./helpers/put')
+    var create = require('./helpers/create')
+
+    var validate = require('./helpers/fuzzing').validate
+
+    var writesPerReplication = ${JSON.stringify(writesPerReplication)}
+
+    tape('autogenerated failing fuzz test', function (t) {
+      create.many(${dbCount}, function (err, dbs, replicateByIndex) {
+        t.error(err)
+        run(${writeOps.map(op => op.toString())})
+      })
+    })`, { singleQuote: true, semi: false }, { encoding: 'utf-8' }, cb))
+}
+
+var testNum = 0
+
+function fuzzRunner (t, opts, cb) {
+  opts = defaultOpts(opts)
+
+  // Used for keeping track of failing test cases.
+  testNum++
+
+  var writesPerReplication = generateData(opts)
+
+  makeDatabases(opts, function (dbs, replicateByIndex) {
+    var ops = []
+    for (var i = 0; i < writesPerReplication.length; i++) {
+      var batch = writesPerReplication[i]
+      var batchOps = []
+      for (var j = 0; j < batch.length; j++) {
+        var singleWrite = batch[j]
+        for (var z = 0; z < singleWrite.values.length; z++) {
+          var value = singleWrite.values[z]
+          if (!value) continue
+          batchOps.push((function (z, k, v) {
+            return eval(`(cb => {
+              put(dbs[${z}], [{
+                key: '${k}',
+                value: '${v}'
+              }], cb)
+            })`)
+          })(z, singleWrite.key, value))
+        }
+      }
+      ops.push(batchOps)
+      // Intersperse replication/validation/failing-test generation between write batches.
+      ops.push([
+        // Currently replicating between all databases at every replication point.
+        cb => replicateByIndex(cb),
+        (function (i) {
+          return eval(`(cb => validate(t, dbs[0], writesPerReplication.slice(0, ${i + 1}), cb))`)
+        })(i),
+      ])
+    }
+
+    var finished = 0
+    doRun()
+
+    function doRun (err) {
+      if (err) {
+        // Don't include the validation/replication ops in the test case generation.
+        // Also don't include the doRun callback in each batch
+        var failingBatches = ops.slice(0, finished).map(batch => batch.slice(0, -1))
+        generateFailingTest(`failing-test-${testNum}`, opts.writers, writesPerReplication, failingBatches, () => {
+          return cb(err)
+        })
```

Comment:
just pass cb directly?

Reply:
Heh was doing more in this cb before. Updated